server/src/main/java/org/elasticsearch/search/rank/RankContextBuilder.java

```diff
@@ -0,0 +1,145 @@
+
+    public static final int DEFAULT_WINDOW_SIZE = SearchService.DEFAULT_SIZE;
+
+    protected int windowSize = DEFAULT_WINDOW_SIZE;
+
+    public RankContextBuilder() {}
+
+    public RankContextBuilder(StreamInput in) throws IOException {
+        windowSize = in.readVInt();
+    }
+
+    public final void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(windowSize);
+        doWriteTo(out);
+    }
+
+    public abstract void doWriteTo(StreamOutput out) throws IOException;
+
+    @Override
+    public final XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.startObject(getWriteableName());
+        builder.field(WINDOW_SIZE_FIELD.getPreferredName(), windowSize);
+        doXContent(builder, params);
+        builder.endObject();
+        builder.endObject();
+        return builder;
+    }
+
+    protected abstract void doXContent(XContentBuilder builder, Params params) throws IOException;
+
+    /**
+     * Allows additional validation as part of {@link SearchRequest#validate} based on subclass rank parameters.
+     */
+    public abstract ActionRequestValidationException validate(
+        ActionRequestValidationException validationException,
+        SearchSourceBuilder source
+    );
+
+    @SuppressWarnings("unchecked")
+    public RCB windowSize(int windowSize) {
+        this.windowSize = windowSize;
+        return (RCB) this;
+    }
+
+    public int windowSize() {
+        return windowSize;
+    }
+
+    /**
+     * Generates the query used for aggregations and suggesters.
+     */
+    public abstract QueryBuilder searchQuery(List<QueryBuilder> queryBuilders);
+
+    /**
+     * Generates a context used to execute required searches on the shard.
+     */
+    public abstract RankShardContext build(List<Query> queries, int size, int from);
+
+    /**
+     * Generates a context used to perform global ranking on the coordinator.
+     */
+    public abstract RankContext build(int size, int from);
+
+    @Override
+    public final boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        @SuppressWarnings("unchecked")
+        RCB other = (RCB) obj;
+        return Objects.equals(windowSize, other.windowSize) && doEquals(other);
+    }
+
+    /**
+     * Indicates whether some other {@link QueryBuilder} object of the same type is "equal to" this one.
+     */
+    protected abstract boolean doEquals(RCB other);
+
+    @Override
+    public final int hashCode() {
+        return Objects.hash(getClass(), windowSize, doHashCode());
+    }
+
+    protected abstract int doHashCode();
+
+    @Override
+    public String toString() {
+        return toString(EMPTY_PARAMS);
+    }
+
+    public String toString(Params params) {
+        try {
```

Comment:
I was using behavior similar to other builders. The nice thing about this is the subclasses get it for "free" since they'll also need they're own x-content methods.

Reply:
doesn't String.toString do the same including handling the exception? My thinking was to further simplify the method, but maybe I am missing why that is not possible

Reply:
Oh, I assumed it was different. Let me double check this.